Switch draw functions from `pydot` to `retworkx`

prakharb10 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Closes #7717 
- [x] `CouplingMap`
- [x] ~`pass_manager_drawer`~
- [x] `Equivalence`


### Details and comments

- Tests for `CouplingMap` do not include any test for the `draw()` function modified here. Should we add a test with a reference image?

prakharb10 commented:
Upon further discussion with @mtreinish, the use of `pydot` in `pass_manager_drawer` cannot be refactored at this point.
Hence, I'm marking the PR ready for review.

mtreinish reviewed (commented):
This LGTM but can you add the tests to add the missing coverage you mentioned in the PR summary? I think we want to just make sure all the functions are exercised before we merge

prakharb10 commented:
I thought about that, but the issue is that the current test class `CouplingTest` or its parent class `QiskitTestCase` do not have any method similar to `QiskitVisualizationTestCase.assertImagesAreEqual`. 
So, should I add a similar method somewhere and then use an image as a reference?

mtreinish commented on test/python/transpiler/test_coupling.py:
```diff
@@ -436,3 +439,9 @@ def test_subgraph(self):
         expected = [(0, 1), (1, 2), (2, 3)]
         self.assertEqual(expected, edge_list, f"{edge_list} does not match {expected}")
+
+    @unittest.skipUnless(optionals.HAS_GRAPHVIZ, "Graphviz not installed")
+    def test_coupling_draw(self):
+        """Test that the coupling map can be drawn."""
+        cmap = CouplingMap([[0, 1], [1, 2], [2, 3], [2, 4], [2, 5], [2, 6]])
+        cmap.draw()
```
Sorry I missed your earlier question. It'd be good if we split this out into a dedicated test class so we can use the image test base class. Then we can assert the output looks like what's expected. Right now all this does is test that creating a coupling map and calling `draw()` doesn't error and it would have passed prior to your fix.

HuangJunye commented:
@prakharb10 Thanks for making this contribution! Are you still working on it? Please let us know if you need any help.

prakharb10 commented:
Hi @HuangJunye. Sorry I was a little busy with college and traveling the last few weeks. By the weekend, I'll add the test case, as @mtreinish said.

HuangJunye commented:
@prakharb10 No problem :D Just want to check in to see if you need any help. Thanks for your response!

prakharb10 commented:
Looks like the coverage report was cancelled. Is this because the base branch is not up to date with `main`?

jakelishman commented:
I've restarted the run. It shouldn't matter if the PR is up-to-date for CI to run. It may have been a remnant of some GitHub Actions downtime late last week.

mtreinish reviewed (commented):
This PR just needs a small tweak to deduplicate all the visualization functionality. I prepared a diff locally, but I think you have allow edits from maintainers unchecked so I can't push it (otherwise I'd just apply it and approve this). So here is the patch file:

```
From 1ef17350d51fda0693d4612e0b24c8b8a2a5cbcd Mon Sep 17 00:00:00 2001
From: Matthew Treinish <mtreinish@kortar.org>
Date: Wed, 31 Aug 2022 17:56:41 -0400
Subject: [PATCH] Deduplicate test definition

---
 test/python/transpiler/test_coupling.py       |  69 +-----------------
 .../references/coupling_map.png               | Bin
 2 files changed, 3 insertions(+), 66 deletions(-)
 rename test/python/{transpiler => visualization}/references/coupling_map.png (100%)

diff --git a/test/python/transpiler/test_coupling.py b/test/python/transpiler/test_coupling.py
index ac061489c..c81916fa2 100644
--- a/test/python/transpiler/test_coupling.py
+++ b/test/python/transpiler/test_coupling.py
@@ -13,10 +13,6 @@
 # pylint: disable=missing-docstring
 
 import unittest
-import os
-from filecmp import cmp as cmpfile
-from shutil import copyfile
-import matplotlib
 
 from qiskit.transpiler import CouplingMap
 from qiskit.transpiler.exceptions import CouplingError
@@ -24,7 +20,7 @@ from qiskit.providers.fake_provider import FakeRueschlikon
 from qiskit.test import QiskitTestCase
 from qiskit.utils import optionals
 
-matplotlib.use("ps")
+from ..visualization.visualization import QiskitVisualizationTestCase, path_to_diagram_reference
 
 
 class CouplingTest(QiskitTestCase):
@@ -447,70 +443,11 @@ class CouplingTest(QiskitTestCase):
         self.assertEqual(expected, edge_list, f"{edge_list} does not match {expected}")
 
 
-def path_to_diagram_reference(filename):
-    return os.path.join(_this_directory(), "references", filename)
-
-
-def _this_directory():
-    return os.path.dirname(os.path.abspath(__file__))
-
-
-class QiskitVisualizationTestCase(QiskitTestCase):
-    """Visual accuracy of visualization tools outputs tests."""
-
-    def assertFilesAreEqual(self, current, expected, encoding="cp437"):
-        """Checks if both file are the same."""
-        self.assertTrue(os.path.exists(current))
-        self.assertTrue(os.path.exists(expected))
-        with open(current, encoding=encoding) as cur, open(expected, encoding=encoding) as exp:
-            self.assertEqual(cur.read(), exp.read())
-
-    def assertEqualToReference(self, result):
-        reference = path_to_diagram_reference(os.path.basename(result))
-        if not os.path.exists(result):
-            raise self.failureException("Result file was not generated.")
-        if not os.path.exists(reference):
-            copyfile(result, reference)
-        if cmpfile(reference, result):
-            os.remove(result)
-        else:
-            raise self.failureException("Result and reference do not match.")
-
-    def assertImagesAreEqual(self, current, expected, diff_tolerance=0.001):
-        """Checks if both images are similar enough to be considered equal.
-        Similarity is controlled by the ```diff_tolerance``` argument."""
-        from PIL import Image, ImageChops
-
-        if isinstance(current, str):
-            current = Image.open(current)
-        if isinstance(expected, str):
-            expected = Image.open(expected)
-
-        diff = ImageChops.difference(expected, current)
-        black_pixels = _get_black_pixels(diff)
-        total_pixels = diff.size[0] * diff.size[1]
-        similarity_ratio = black_pixels / total_pixels
-        self.assertTrue(
-            1 - similarity_ratio < diff_tolerance,
-            f"The images are different by {(1 - similarity_ratio) * 100}%"
-            f" which is more than the allowed {diff_tolerance * 100}%",
-        )
-
-
-def _get_black_pixels(image):
-    black_and_white_version = image.convert("1")
-    black_pixels = black_and_white_version.histogram()[0]
-    return black_pixels
-
-
 class CouplingVisualizationTest(QiskitVisualizationTestCase):
-    def setUp(self):
-        super().setUp()
-        self.cmap = CouplingMap([[0, 1], [1, 2], [2, 3], [2, 4], [2, 5], [2, 6]])
-
     @unittest.skipUnless(optionals.HAS_GRAPHVIZ, "Graphviz not installed")
     def test_coupling_draw(self):
         """Test that the coupling map drawing with respect to the reference file is correct."""
+        cmap = CouplingMap([[0, 1], [1, 2], [2, 3], [2, 4], [2, 5], [2, 6]])
         image_ref = path_to_diagram_reference("coupling_map.png")
-        image = self.cmap.draw()
+        image = cmap.draw()
         self.assertImagesAreEqual(image, image_ref, diff_tolerance=0.01)
diff --git a/test/python/transpiler/references/coupling_map.png b/test/python/visualization/references/coupling_map.png
similarity index 100%
rename from test/python/transpiler/references/coupling_map.png
rename to test/python/visualization/references/coupling_map.png
-- 
2.37.2
```

After you apply that diff to your branch this should be good to go.

prakharb10 commented:
Sorry about that, @mtreinish. Edits from maintainers are actually ON. I have no idea why it didn't allow you to make the changes. Anyway, I've made the changes you requested.
I've also added a similar visualization test for the `EquivalenceLibrary.draw()` function, which was also switched to `rustworkx` and didn't have any tests before.
